Flaky test: test_peak_memory_usage/test.py::test_clickhouse_client_max_peak_memory_usage_distributed

### Highlights
- https://github.com/ClickHouse/ClickHouse/issues/90840#issuecomment-3662483887

### Report

Failure reason: assert shard_2.contains_in_log(f"Query peak memory usage: {peak_memory_usage}")
[CI report: [Integration tests (amd_binary, 2/5)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=90717&sha=2d993904c4f95324ba1379a2c350943ce038502c&name_0=PR&name_1=Integration%20tests%20%28amd_binary%2C%202%2F5%29)]
[cidb](https://play.clickhouse.com/play?user=play&run=1#V0lUSAogICAgOTAgQVMgaW50ZXJ2YWxfZGF5cwpTRUxFQ1QKICAgIHRvU3RhcnRPZkRheShjaGVja19zdGFydF90aW1lKSBBUyBkYXksCiAgICBjb3VudCgpIEFTIGZhaWx1cmVzLAogICAgZ3JvdXBVbmlxQXJyYXkocHVsbF9yZXF1ZXN0X251bWJlcikgQVMgcHJzLAogICAgYW55KHJlcG9ydF91cmwpIEFTIHJlcG9ydF91cmwKRlJPTSBjaGVja3MKV0hFUkUgKG5vdygpIC0gdG9JbnRlcnZhbERheShpbnRlcnZhbF9kYXlzKSkgPD0gY2hlY2tfc3RhcnRfdGltZQogICAgQU5EIHRlc3RfbmFtZSA9ICd0ZXN0X3BlYWtfbWVtb3J5X3VzYWdlL3Rlc3QucHk6OnRlc3RfY2xpY2tob3VzZV9jbGllbnRfbWF4X3BlYWtfbWVtb3J5X3VzYWdlX2Rpc3RyaWJ1dGVkJwogICAgLS0gQU5EIGNoZWNrX25hbWUgPSAnSW50ZWdyYXRpb24gdGVzdHMgKGFtZF9iaW5hcnksIDIvNSknCiAgICBBTkQgdGVzdF9zdGF0dXMgSU4gKCdGQUlMJywgJ0VSUk9SJykKICAgIEFORCAoKGhlYWRfcmVmID0gJ21hc3RlcicgQU5EIHB1bGxfcmVxdWVzdF9udW1iZXIgPSAwKSBPUiBwdWxsX3JlcXVlc3RfbnVtYmVyICE9IDApCkdST1VQIEJZIGRheQpPUkRFUiBCWSBkYXkgREVTQwo=)

Test output:
```
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
FAIL [test_peak_memory_usage/test.py::test_clickhouse_client_max_peak_memory_usage_distributed]
  | File: test_peak_memory_usage/test.py:101 - in test_clickhouse_client_max_peak_memory_usage_distributed
  |     assert shard_2.contains_in_log(f"Query peak memory usage: {peak_memory_usage}")
  | E   AssertionError: assert False
  | E    +  where False = contains_in_log('Query peak memory usage: 8.23')
  | E    +    where contains_in_log = <helpers.cluster.ClickHouseInstance object at 0x7f62ad427dc0>.contains_in_log
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

```

--- a/src/Server/TCPHandler.cpp
+++ b/src/Server/TCPHandler.cpp
@@ -800,12 +800,12 @@ void TCPHandler::runImpl()
             {
                 /// FIXME: check explicitly that insert query suggests to receive data via native protocol,
                 query_state->need_receive_data_for_insert = true;
-                processInsertQuery(*query_state);
+                processInsertQuery(*query_state, *query_scope);
                 query_state->io.onFinish();
             }
             else if (query_state->io.pipeline.pulling())
             {
-                processOrdinaryQuery(*query_state);
+                processOrdinaryQuery(*query_state, *query_scope);
                 query_state->io.onFinish();
             }
             else if (query_state->io.pipeline.completed())
@@ -830,7 +830,12 @@ void TCPHandler::runImpl()
                 /// NOTE: we cannot send Progress for regular INSERT (with VALUES)
                 /// without breaking protocol compatibility, but it can be done
                 /// by increasing revision.
+
                 sendProgress(*query_state);
+
+                /// Log peak memory usage just before sending it to client to make it as accurate as possible
+                /// (though note we may still have some allocations in between, that will make the difference)
+                query_scope->logPeakMemoryUsage();
                 sendSelectProfileEvents(*query_state);
             }
             else
@@ -843,13 +848,13 @@ void TCPHandler::runImpl()
                     create_query && create_query->isCreateQueryWithImmediateInsertSelect())
                 {
                     sendProgress(*query_state);
+
+                    /// Log peak memory usage just before sending it to client to make it as accurate as possible
+                    /// (though note we may still have some allocations in between, that will make the difference)
+                    query_scope->logPeakMemoryUsage();
                     sendSelectProfileEvents(*query_state);
                 }
             }
-
-            /// Do it before sending end of stream, to have a chance to show log message in client.
-            query_scope->logPeakMemoryUsage();
-
             sendLogs(*query_state);
             sendEndOfStream(*query_state);
 
@@ -1261,7 +1266,7 @@ AsynchronousInsertQueue::PushResult TCPHandler::processAsyncInsertQuery(QuerySta
 }
 
 
-void TCPHandler::processInsertQuery(QueryState & state)
+void TCPHandler::processInsertQuery(QueryState & state, CurrentThread::QueryScope & query_scope)
 {
     size_t num_threads = state.io.pipeline.getNumThreads();
 
@@ -1350,11 +1355,14 @@ void TCPHandler::processInsertQuery(QueryState & state)
         run_executor(executor, std::move(processed_block));
     }
 
+    /// Log peak memory usage just before sending it to client to make it as accurate as possible
+    /// (though note we may still have some allocations in between, that will make the difference)
+    query_scope.logPeakMemoryUsage();
     sendInsertProfileEvents(state);
 }
 
 
-void TCPHandler::processOrdinaryQuery(QueryState & state)
+void TCPHandler::processOrdinaryQuery(QueryState & state, CurrentThread::QueryScope & query_scope)
 {
     auto & pipeline = state.io.pipeline;
 
@@ -1436,6 +1444,10 @@ void TCPHandler::processOrdinaryQuery(QueryState & state)
         sendProfileInfo(state, executor.getProfileInfo());
         sendProgress(state);
         sendLogs(state);
+
+        /// Log peak memory usage just before sending it to client to make it as accurate as possible
+        /// (though note we may still have some allocations in between, that will make the difference)
+        query_scope.logPeakMemoryUsage();
         sendSelectProfileEvents(state);
 
         sendData(state, {});

--- a/src/Server/TCPHandler.h
+++ b/src/Server/TCPHandler.h
@@ -301,11 +301,11 @@ private:
 
     /// Process INSERT query
     void startInsertQuery(QueryState & state);
-    void processInsertQuery(QueryState & state);
+    void processInsertQuery(QueryState & state, CurrentThread::QueryScope & query_scope);
     AsynchronousInsertQueue::PushResult processAsyncInsertQuery(QueryState & state, AsynchronousInsertQueue & insert_queue);
 
     /// Process a request that does not require the receiving of data blocks from the client
-    void processOrdinaryQuery(QueryState & state);
+    void processOrdinaryQuery(QueryState & state, CurrentThread::QueryScope & query_scope);
 
     void processTablesStatusRequest();
 

